controller/controller.go

```diff
@@ -1114,25 +1114,61 @@ func (ctrl *ProvisionController) syncVolume(ctx context.Context, obj interface{}
 }
 
 func (ctrl *ProvisionController) handleProtectionFinalizer(ctx context.Context, volume *v1.PersistentVolume) (*v1.PersistentVolume, error) {
-	var modifiedFinalizers []string
+	//var modifiedFinalizers []string
 	var modified bool
+	klog.Infof("handleProtectionFinalizer Volume : %+v", volume)
 	reclaimPolicy := volume.Spec.PersistentVolumeReclaimPolicy
+	volumeFinalizers := volume.ObjectMeta.Finalizers
+	if metav1.HasAnnotation(volume.ObjectMeta, annDynamicallyProvisioned) {
+		provisionPluginName := volume.Annotations[annDynamicallyProvisioned]
+		migratedAnn := volume.Annotations[annMigratedTo]
+		// Determine if the PV is owned by the current provisioner.
+		if !ctrl.knownProvisioner(provisionPluginName) && !ctrl.knownProvisioner(migratedAnn) {
+			// The current provisioner is not responsible for adding the finalizer
+			return volume, nil
+		}
+		// Add the finalizer only if `addFinalizer` config option is enabled, finalizer doesn't exist and PV is not already
+		// under deletion.
+		if ctrl.addFinalizer && reclaimPolicy == v1.PersistentVolumeReclaimDelete && volume.DeletionTimestamp == nil {
+			volumeFinalizers, modified = addFinalizer(volumeFinalizers, finalizerPV)
+		}
+	} else {
+		// Statically provisioned volume. Check if the volume type is CSI.
+		if volume.Spec.PersistentVolumeSource.CSI != nil {
+			volumeProvisioner := volume.Spec.CSI.Driver
+			if !ctrl.knownProvisioner(volumeProvisioner) {
+				return volume, nil
+			}
+		} else {
+			// Check if the volume is being migrated
+			migratedAnn := volume.Annotations[annMigratedTo]
+			if !ctrl.knownProvisioner(migratedAnn) {
+				// The current provisioner is not responsible for adding the finalizer
+				return volume, nil
+			}
+		}
+		// Add the finalizer only if `addFinalizer` config option is enabled, finalizer doesn't exist and PV is not already
+		// under deletion and the volume is in a Bound state.
+		if ctrl.addFinalizer && reclaimPolicy == v1.PersistentVolumeReclaimDelete && volume.DeletionTimestamp == nil && volume.Status.Phase == v1.VolumeBound {
+			volumeFinalizers, modified = addFinalizer(volumeFinalizers, finalizerPV)
+		}
+		// For statically provisioned volumes remove the external finalizer if the volume is not in a Bound state
+		if volume.Status.Phase != v1.VolumeBound {
```

Comment:
This will remove the finalizer from PVs that get into Released state when the provisioner should make sure the volume is deleted. The provisioner should *keep* the finalizer and keep deleting the volume until it succeeds.

IMO it's enough to *add* the finalizer when  `Phase == v1.VolumeBound` (1 line above) and remove it after deletion succeeds.

The KEP says:

> Statically provisioned volumes would behave the same as dynamically provisioned volumes except in cases where the PV is not associated with a PVC, in such cases finalizer external-provisioner.volume.kubernetes.io/finalizer is not added.

Reply:
The existing code triggers deletion from the backend only for dynamically provisioned volumes https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/controller.go#L1227

This will need changes there as well, which makes sense. Our goal is to ensure that volume gets deleted from backend irrespective of if it was statically or dynamically provisioned. Will make these changes.